tide/wall/SVGTiler.h

```diff
@@ -42,26 +42,30 @@
 
 #include "LodTiler.h"
 
-#include "SVGTextureFactory.h"
+#include "SVG.h"
 
 /**
- * Reprensent an SVG image as a multi-LOD tiled data source.
+ * Represent an SVG image as a multi-LOD tiled data source.
  */
 class SVGTiler : public LodTiler
 {
 public:
     /** Constructor. */
-    explicit SVGTiler( SVGTextureFactoryPtr factory );
+    explicit SVGTiler( SVG& svg );
 
 private:
     /**
      * Get a tile image which will be cached.
-     * Unlike other DataSource classes, this method must be called from a thread
-     * with an OpenGL context.
+     * Unlike other DataSource classes, this method may need to be called from a
+     * thread with an OpenGL context depending on the SVG backend used.
      */
     QImage getCachableTileImage( uint tileId ) const final;
 
-    SVGTextureFactoryPtr _factory;
+    SVG& _svg;
+
+    mutable QMutex _threadMapMutex;
+    typedef std::unique_ptr<SVG> SVGPtr;
+    mutable std::map<Qt::HANDLE, SVGPtr> _perThreadSVG;
```

Comment:
Interesting point, but there is a catch here. Apparently QThreadStorage destructor does not deallocate objects, they are instead deleted when the thread that created them exits. Since the calling threads are from the global QThreadPool, in our case the resources would be "leaked" until the application exits. boost::thread_specific_ptr has the same behaviour, and is meant to be used with boost::thread instead of QThread. The reason invoked for this is that it may not be safe to deallocate objects allocated by a different thread. I am not sure if this may be a concern in this case, I copied this implementation from the PDFTiler where we were doing the same.

In any case this multi-threading is quite problematic because allocating a new RSVG context for a ~80MB file takes in the order of ~5s, but then subsequent render calls are faster because they are fully parallel. Using a single RSVG context, the initial impact is lower, but then the render call have to be serialized because RSVG is not thread-safe, which can even slow down the entire QThreadPool. I am not sure what is the optimal solution here.